internal/cmd/local/k8s/load_images.go

```diff
@@ -0,0 +1,208 @@
+
+		go func(img string) {
+			defer wg.Done()
+			r, err := dockerClient.ImagePull(ctx, img, image.PullOptions{})
+			if err != nil {
+				pterm.Debug.Printfln("error pulling image %s", err)
+				// image pull errors are intentionally dropped because we're in a goroutine,
+				// and because we don't want to interrupt other image pulls.
+			}
+			defer r.Close()
+			io.Copy(io.Discard, r)
+		}(img)
+	}
+	wg.Wait()
+
+	// The context could be canceled by now. If so, return early.
+	if ctx.Err() != nil {
+		return ctx.Err()
+	}
+
+	// Determine which images need to be loaded onto the nodes.
+	needed := determineImagesForLoading(ctx, dockerClient, images, nodes)
+	if len(needed) == 0 {
+		return nil
+	}
+
+	// for _, node := range nodes {
+	// 	for _, img := range needed {
+
+	// 		pterm.Debug.Printfln("saving image %s into node %s", img, node)
+
+	// 		r, err := dockerClient.ImageSave(ctx, images)
+	// 		if err != nil {
+	// 			pterm.Debug.Printfln("error loading image %s into node %s: %s", img, node, err)
+	// 			continue
+	// 		}
+
+	// 		pterm.Debug.Printfln("loading image %s into node %s", img, node)
+
+	// 		err = nodeutils.LoadImageArchive(node, r)
+	// 		if err != nil {
+	// 			pterm.Debug.Printfln("error loading image %s into node %s: %s", img, node, err)
+	// 		}
+	// 	}
+	// }
+
+	// Save all the images to an archive, images.tar
+	imagesTarPath, err := saveImageArchive(ctx, dockerClient, needed)
+	if err != nil {
+		return fmt.Errorf("failed to save image archive: %w", err)
+	}
+	// defer os.RemoveAll(imagesTarPath)
+
+	// Load the image archive into the Kind nodes.
+	f, err := os.Open(imagesTarPath)
+	if err != nil {
+		return fmt.Errorf("failed to open image archive: %w", err)
+	}
+	defer f.Close()
+
+	for _, n := range nodes {
+		pterm.Debug.Printfln("loading image archive into kind node %s", n)
+		err := nodeutils.LoadImageArchive(n, f)
+		if err != nil {
+			pterm.Debug.Printfln("%s", err)
+		}
+	}
+	return nil
+}
+
+// getExistingImageDigests returns the set of images that already exist on the nodes.
+func getExistingImageDigests(ctx context.Context, nodes []nodeslib.Node) common.Set[string] {
+	existingByNode := map[string]int{}
+
+	for _, n := range nodes {
+
+		out, err := exec.CombinedOutputLines(n.CommandContext(ctx, "ctr", "--namespace=k8s.io", "images", "list"))
+		if err != nil {
+			// ignore the error because discovering these images is just an optimization.
+			pterm.Debug.Printfln("error discovering existing images: %s %s", err, out)
+			continue
+		}
+		if len(out) < 1 {
+			continue
+		}
+
+		// the first line is a header. verify the columns we expect, just in case the format ever changes.
+		header := strings.Fields(out[0])
+		if len(header) < 1 || header[0] != "REF" {
+			pterm.Debug.Printfln("unexpected format from ctr image list. skipping node %s.", n)
+			continue
+		}
+
+		// skip the first line, which is a header.
+		for _, l := range out[1:] {
+			ref := strings.Fields(l)[0]
```

Comment:
Any chance of this blowing up?  I haven't looked at what `Fields` returns.

Reply:
I do validate the header above this, but I added an extra defensive check anyway